### PR TITLE
ENH: Improve performance for df.__setitem__ with list-like indexers

### LINKED ISSUE #37954: BUG: df.__setitem__ can be 10x slower than pd.concat(..., axis=1)
- [x] I have checked that this issue has not already been reported.
  * I think so, it's a bit hard to search for though. Nothing under concat and __setindex__

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import pandas as pd, numpy as np
from string import ascii_lowercase

def setitem(x, x_cols, df):
    new = pd.DataFrame(index=df.index)
    new[x_cols] = x
    new[df.columns] = df
    return new


def concat(x, x_cols, df):
    return pd.concat(
        [
            pd.DataFrame(x, columns=x_cols, index=df.index),
            df,
        ],
        axis=1,
    )

x = np.ones((1000, 10))
x_col = list(ascii_lowercase[:10])
df = pd.DataFrame(
    {
        "str": np.random.choice(np.array(list(ascii_lowercase)), size=1000),
        "int": np.arange(1000, dtype=int),
    }
)


%timeit setitem(x, x_col, df)
# 3.78 ms ± 193 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%timeit concat(x, x_col, df)
# 306 µs ± 9.29 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

#### Problem description

This seems unintuitive from a performance perspective. I would assume that these would be close to equivalent. The `setitem` implementation might even be expected to have better performance due to less allocation.

While this is a stripped down example, the use case is building a dataframe to return from a function. Making a dataframe, then adding columns seemed like the natural idiom here.

#### Output of ``pd.show_versions()``

<details>

```
INSTALLED VERSIONS
------------------
commit           : 67a3d4241ab84419856b84fc3ebc9abcbe66c6b3
python           : 3.8.5.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 19.6.0
Version          : Darwin Kernel Version 19.6.0: Thu Oct 29 22:56:45 PDT 2020; root:xnu-6153.141.2.2~1/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.1.4
numpy            : 1.19.4
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.2.4
setuptools       : 50.3.2
Cython           : 0.29.21
pytest           : 6.1.2
hypothesis       : None
sphinx           : 3.2.1
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.19.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : 0.8.4
fastparquet      : None
gcsfs            : None
matplotlib       : 3.3.3
numexpr          : 2.7.1
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 2.0.0
pytables         : None
pyxlsb           : None
s3fs             : None
scipy            : 1.5.4
sqlalchemy       : 1.3.18
tables           : 3.6.1
tabulate         : 0.8.7
xarray           : 0.16.1
xlrd             : 1.2.0
xlwt             : None
numba            : 0.51.2
```

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #37954
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Reindexing the Block Manager improves the performance significantly. I hope I have not missed anything, concerning the reindexing of the blocks. 
Time spent in ``_ensure_listlike_indexer`` is pretty low now.
timeit result for the ops methods:

```
In [20]: %timeit setitem(x, x_col, df)
1.09 ms ± 9.65 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [21]: %timeit concat(x, x_col, df)
293 µs ± 4.29 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Should we add tests here? I have added an asv to capture this case.

cc @jbrockmendel

### CODE REVIEW

File: asv_bench/benchmarks/indexing.py [hunk 1]
```diff
@@ -358,6 +358,10 @@ def time_assign_with_setitem(self):
         for i in range(100):
             self.df[i] = np.random.randn(self.N)
 
+    def time_assign_list_like_with_setitem(self):
```
Comment (jreback) [inline asv_bench/benchmarks/indexing.py:361]: do we also have the concat example (you can put it here)?

File: pandas/core/indexing.py [hunk 2]
```diff
@@ -663,17 +663,9 @@ def _ensure_listlike_indexer(self, key, axis=None, value=None):
...
-                    if value is None:
-                        self.obj[k] = np.nan
-                    elif is_array_like(value) and value.ndim == 2:
-                        # GH#37964 have to select columnwise in case of array
-                        self.obj[k] = value[:, i]
-                    elif is_list_like(value):
-                        self.obj[k] = value[i]
-                    else:
-                        self.obj[k] = value
+            keys = self.obj.columns.tolist()
```
Comment (jreback) [inline pandas/core/indexing.py:666]: can you create the keys in the same order whether they are in the obj or not (e.g. combine L666 and 667)

Comment (jreback) [reply pandas/core/indexing.py:666]: do we have a test for this ordering?

**Review by jreback — CHANGES_REQUESTED** [review]
wow, simplification and pref boost

File: doc/source/whatsnew/v1.2.0.rst [hunk 3]
```diff
@@ -257,6 +257,7 @@ Other enhancements
 - Improve numerical stability for :meth:`.Rolling.skew`, :meth:`.Rolling.kurt`, :meth:`Expanding.skew` and :meth:`Expanding.kurt` through implementation of Kahan summation (:issue:`6929`)
 - Improved error reporting for subsetting columns of a :class:`.DataFrameGroupBy` with ``axis=1`` (:issue:`37725`)
 - Implement method ``cross`` for :meth:`DataFrame.merge` and :meth:`DataFrame.join` (:issue:`5401`)
+- Improve performance for :meth:`DataFrame.__setitem__` with list-like indexers (:issue:`37954`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.2.0.rst:260]: i think it would be ok to backport this to 1.1.5

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexing.py [hunk 4]
```diff
@@ -663,17 +663,9 @@ def _ensure_listlike_indexer(self, key, axis=None, value=None):
...
+            keys.extend([k for k in key if k not in self.obj])
```
Comment (jbrockmendel) [inline pandas/core/indexing.py:667]: why are you doing this with a list instead of Index.union?

Comment (phofl) [reply pandas/core/indexing.py:667]: Haven't thoug about union. Thanks. Changed it.

File: doc/source/whatsnew/v1.2.0.rst (hunk 3 above)
Comment (phofl) [reply doc/source/whatsnew/v1.2.0.rst:260]: Moved to 1.1.5

File: asv_bench/benchmarks/indexing.py (hunk 1 above)
Comment (phofl) [reply asv_bench/benchmarks/indexing.py:361]: Done

File: pandas/core/indexing.py (hunk 2 above)
Comment (phofl) [reply pandas/core/indexing.py:666]: Union should do the trick.

At least the test added with #37964 covers this

File: pandas/core/indexing.py (hunk 4 above)
Comment (phofl) [reply pandas/core/indexing.py]: Gave us another slight performance improvement

```
In [16]: %timeit setitem(x, x_col, df)
1.03 ms ± 9.96 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm can merge on green.

**Comment by phofl** [comment]
@jreback greenish. Failure unrelated

**Comment by jreback** [comment]
thanks @phofl

**Comment by jbrockmendel** [comment]
@phofl did you determine that this isn't making a copy?

**Comment by phofl** [comment]
Not with a test, but I tested this in code.
```
x = self.obj
self.obj._mgr = self.obj._mgr.reindex_axis(keys, 0)
```

This also changes x, so that is not a copy, isn't it?

**Comment by jbrockmendel** [comment]
> This also changes x, so that is not a copy, isn't it?

Assuming homogeneous dtype for now, what im asking for is:

```
values = self.obj.values
self.obj._mgr = self.obj._mgr.reindex_axis(keys, 0)
new_values = self.obj.values

assert new_values is values   # <-- will raise if we made a copy
```

**Comment by jreback** [comment]
i am not sure this is possible w/o a copy

**Comment by phofl** [comment]
This does make a copy. Did not know, that we have to compare values instead of the object itself.

**Comment by phofl** [comment]
Actually this raised previously too, so at least not a regression

**Comment by jbrockmendel** [comment]
> Did not know, that we have to compare values instead of the object itself.

Note: the example above I assumed homogeneous dtypes.  More generally, we need to have `obj._mgr.blocks[n].values` be unchanged for each pre-existing n.

> i am not sure this is possible w/o a copy

A few things need to happen for it to be possible:

1) add `consolidate: bool = True` keyword to `BlockManager.reindex_axis` that then passes consolidate=consolidate to `reindex_indexer`
2) add `only_slice: bool = False` to `reindex_axis` and `reindex_indexer` so it can be passed to `_slice_take_blocks_ax0`
3) the `reindex_axis` call in this PR passes `consolidate=False, only_slice=True`

**Comment by simonjayhawkins** [comment]
@meeseeksdev backport 1.1.x

**Comment by simonjayhawkins** [comment]
https://github.com/pandas-dev/pandas/pull/38181#issuecomment-735682557

> @phofl #38148 (comment) was this necessary? do we know what commit caused the performance regression. the existing code on 1.1.x looks much simpler to what was replaced on master.

If we don't backport this, we will need to move the release note on master

**Comment by jorisvandenbossche** [comment]
As @jbrockmendel noted, this *can* change the copy/view semantics in certain cases. 

One example:

```python
df = pd.DataFrame({'a': [1, 2, 3], 'b': [4, 4, 6]})
# get one column as a view of df
s = df['a']
# add columns with list-like indexer
df[['c', 'd']] = np.array([[.1, .2], [.3, .4], [.4, .5]])
# edit in place the first column to check view semantics
df.iloc[0, 0] = 100
```

on master this gives:
```
In [8]: df
Out[8]: 
     a  b    c    d
0  100  4  0.1  0.2
1    2  4  0.3  0.4
2    3  6  0.4  0.5

In [9]: s
Out[9]: 
0    1
1    2
2    3
Name: a, dtype: int64
```

where "a" and "b" columns were now copied in the `df[['c', 'd']] = ...` operation (shown by `s` not being updated). 

While on pandas 1.1.4, this copy didn't happen, and the series was actually updated:

```
In [6]: df
Out[6]: 
     a  b    c    d
0  100  4  0.1  0.2
1    2  4  0.3  0.4
2    3  6  0.4  0.5

In [7]: s
Out[7]: 
0    100
1      2
2      3
Name: a, dtype: int64
```

This of course closely relates to the recent discussions about improving the copy/view semantics (as currently those semantics are both not clear and largely untested).

(now the above is certainly a specific example to trigger the issue. In many cases we would also do a copy, eg if there would already have been float columns present in the 
Given those ungoing discussions, we maybe shouldn't change those semantics now (I think the goal of of the discussions would certainly be

**Comment by jbrockmendel** [comment]
what if instead of calling reindex_axis we called reindex_indexer and passed `indexer=slice(len(self.obj.columns))`?

**Pull request merged**